### ISSUE TITLE
mat-stepper: Allow custom CSS classes for mat-step-header

### ISSUE THREAD

**Issue by klinki** [angular/material2#15478]
#### Please describe the feature you would like to request.
Please add API to add custom css classes to `mat-step-header`.

#### What is the use-case or motivation for this proposal?
We want to set custom cursor for disabled steps in linear stepper.

#### Is there anything else we should know?
Disabled steps have the same hover and click animations, it would be nice to be able to customize them. But I guess there is other issue for that. And custom css classes would also help with this.

**Comment by hriverahdez** [comment]
Yes please add this feature. I would like to apply styles to the active step's header via a class.
I've done it like this, but is kind of a hack

`.mat-step-header[aria-selected="true"] {
	background-color: rgba(0, 0, 0, 0.04) !important
}`

**Comment by joebochill** [comment]
I too would like to be able to style my step headers via CSS classes. Currently resorting to ```::ng-deep``` rules which I am not a fan of.

**Comment by tinybigideas** [comment]
This would be very welcomed. Currently using a CSS workaround to disable a couple optional steps dynamically:

```
#cdk-step-label-0-1,
#cdk-step-label-0-2 {
  pointer-events: none !important;
  cursor: not-allowed;
  opacity: 0.6;
}
```

Not ideal.

**Comment by cornyluft** [comment]
I too would like to be able to style the step headers via CSS classes.

**Comment by davideas** [comment]
Well, not only the step-headers, but the single step `<mat-step>` !!!
Currently it changes the element name at runtime, ignoring the id and the added class names, so we can't intercept the single element via CSS.
Give us an option to identify the elements via its ID or our class name.

**Comment by andreElrico** [comment]
I found a way how to disable a step. The material team should add a disable option on `<mat-step ...>`

here is a working [STACKBLITZ](https://stackblitz.com/edit/angular-9jze9y?file=src%2Fapp%2Fstepper-states-example.css)

I (miss)use the `aria-labelledby` input.

```
...
<mat-step [stepControl]="secondFormGroup" [aria-labelledby]="stepDisabled ? 'disabled_af' : null">
...
```

in css:

```
::ng-deep .mat-step-header[aria-labelledby="disabled_af"] {
  pointer-events: none !important;
  cursor: not-allowed;
  opacity: 0.6;
  background-color: pink /* remove when in prod. Its just there to prove the concept*/
}
```

**Comment by FloNeu** [comment]
Related - i just tried to style the 'stepper-lines' and other parts of the mat-step component, depending on the state of the mat-step, like 'error', 'completed', 'editable'. I tried to apply These styles to the <mat-step>-container using [ngclass] which doesn't work - i think it would be immensely useful to allow a way to apply custom classes ( or at least these specific stepper-state classes to the  mat-step container ).

**Comment by kemalk89** [comment]
In our case all steps are completed / done. But the selected step icon AND all the step icons before should have a specific background-color. The remaining step icons should have a different color. The idea is to highlight visited and current step icons. So the ability to inject own classnames would be great. :-)

**Comment by willherr** [comment]
> I found a way how to disable a step. The material team should add a disable option on `<mat-step ...>`
> 
> here is a working [STACKBLITZ](https://stackblitz.com/edit/angular-9jze9y?file=src%2Fapp%2Fstepper-states-example.css)
> 
> I (miss)use the `aria-labelledby` input.
> 
> ```
> ...
> <mat-step [stepControl]="secondFormGroup" [aria-labelledby]="stepDisabled ? 'disabled_af' : null">
> ...
> ```
> 
> in css:
> 
> ```
> ::ng-deep .mat-step-header[aria-labelledby="disabled_af"] {
>   pointer-events: none !important;
>   cursor: not-allowed;
>   opacity: 0.6;
>   background-color: pink /* remove when in prod. Its just there to prove the concept*/
> }
> ```
> 
> * please use some other aria label as I have here.
> * of course you have to manage the disable-state of the stepper buttons with the known techniques.

What a terrible workaround...that I will be using.

Thanks!

**Comment by Nxtivision** [comment]
There is another way to do this without interacting with accessibility 

For example, if I want to disable one of my mat-step-header:
```html
<mat-step extendedStep label="Step 2" [disabled]="true">
    Step disabled
</mat-step>
```

And my directive:

```typescript
@Directive({
  selector: '[extendedStep]',
  exportAs: 'extendedStep',
})
export class ExtendedStepDirective implements OnChanges, AfterViewInit {
  @Input()
  disabled = false;
  private readonly DISABLED_CLASS = 'mat-step-header-disabled';

  private stepHeader!: MatStepHeader;
  private renderer = inject(Renderer2);

  constructor(private step: MatStep) {}

  ngAfterViewInit() {
    const stepper = this.step._stepper;
    const indexStep = stepper.steps.toArray().indexOf(this.step);
    this.stepHeader = stepper._stepHeader.get(indexStep) as MatStepHeader;
    if (this.disabled) this.addClass(this.DISABLED_CLASS);
  }

  ngOnChanges(changes: SimpleChanges) {
    if (!changes['disabled'].isFirstChange()) {
      changes['disabled'].currentValue
        ? this.addClass(this.DISABLED_CLASS)
        : this.removeClass(this.DISABLED_CLASS);
    }
  }

  private addClass(className: string) {
    this.renderer.addClass(
      this.stepHeader._elementRef.nativeElement,
      className
    );
  }

  private removeClass(className: string) {
    this.renderer.removeClass(
      this.stepHeader._elementRef.nativeElement,
      className
    );
  }
}
```

Of course, if we have multiple inputs, it could be interesting to refactor how we manipulate inputs inside a method. For my case, it's useful 😃 

Hope this will help
